components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/StorageMaintenanceWorker.kt

```diff
@@ -0,0 +1,48 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.storage.sync
+
+import android.content.Context
+import androidx.work.CoroutineWorker
+import androidx.work.Data
+import androidx.work.WorkerParameters
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+
+/**
+ * WorkManager's Worker that runs maintenance task on [PlacesStorage].
+ * When [doWork] returns with a failure or Worker's constraints set via
+ * [PlacesHistoryStorage.getStorageWorkerConstraints] are not met anymore, i.e. catches an
+ * exception while executing its task, write operations on [PlacesStorage] are cancelled.
+ * */
+internal class StorageMaintenanceWorker(context: Context, params: WorkerParameters) :
+    CoroutineWorker(context, params) {
+
+    @Suppress("TooGenericExceptionCaught")
+    override suspend fun doWork(): Result =
+        try {
+            withContext(Dispatchers.IO) {
+                GlobalPlacesDependencyProvider.requirePlacesStorage()
+                    .runMaintenance(DB_SIZE_LIMIT_IN_BYTES)
+                Result.success()
+            }
+        } catch (exception: Exception) {
+            withContext(Dispatchers.IO) {
+                GlobalPlacesDependencyProvider.requirePlacesStorage()
+                    .cancelWrites()
+                Result.failure(
+                    Data.Builder().putString(STORAGE_MAINTENANCE_WORKER_TAG, exception.message)
+                        .build(),
```

Comment:
we are not doing anything with the data; I included that case only in the test cases to validate that I am returning the message when there is an exception, in case it is decided to be used.

Reply:
Ah yes, I missed that assertion in the test. I think that's fine to have then.